### PR TITLE
gh-105156: Cleanup usage of old Py_UNICODE type

### PR DESCRIPTION
**Pull request by vstinner**
* refcounts.dat:

  * Remove Py_UNICODE functions
  * Replace Py_UNICODE argument type with wchar_t

* _PyUnicode_ToLowercase(), _PyUnicode_ToUppercase(), _PyUnicode_ToTitlecase() are no longer deprecate in comment. It's no longer needed since they now use Py_UCS4 type, rather than the deprecated Py_UNICODE type.
* gdb: Remove unused char_width() method.

* Issue: gh-105156

----
:books: Documentation preview :books:: https://cpython-previews--105158.org.readthedocs.build/

### CODE REVIEW

**Comment by vstinner** [comment]
cc @methane

File: Tools/gdb/libpython.py [hunk 1]
```diff
@@ -1414,13 +1410,13 @@ def proxyval(self, visited):
         elif repr_kind == 4:
             field_str = field_str.cast(_type_unsigned_int_ptr())
 
-        # Gather a list of ints from the Py_UNICODE array; these are either
+        # Gather a list of ints from the character array; these are either
```
Comment (vstinner) [inline Tools/gdb/libpython.py]: I hesitate to say "code points" here.

Comment (methane) [reply Tools/gdb/libpython.py]: We use "code points" in next line. I think "code points" is better.

File: Tools/gdb/libpython.py [hunk 2]
```diff
@@ -1414,13 +1410,13 @@ def proxyval(self, visited):
...
         # UCS-1, UCS-2 or UCS-4 code points:
-        Py_UNICODEs = [int(field_str[i]) for i in safe_range(field_length)]
+        characters = [int(field_str[i]) for i in safe_range(field_length)]
 
         # Convert the int code points to unicode characters, and generate a
         # local unicode instance.
-        result = u''.join(map(chr, Py_UNICODEs))
+        result = u''.join(map(chr, characters))
```
Comment (methane) [inline Tools/gdb/libpython.py]:
```suggestion
        result = ''.join(map(chr, characters))
```

File: Tools/gdb/libpython.py (hunk 1 above)
Comment (vstinner) [reply Tools/gdb/libpython.py]: Alright, I changed it to code points.

File: Tools/gdb/libpython.py (hunk 2 above)
Comment (vstinner) [reply Tools/gdb/libpython.py]: Removing supporting for Python 2 require way more changes. I prefer to restrict changes to just Py_UNICODE here.

**Comment by vstinner** [comment]
Merged. Thanks for your review @methane.

File: Tools/gdb/libpython.py (hunk 2 above)
Comment (methane) [reply Tools/gdb/libpython.py]: FYI, I removed Python 2 support from libpython.py already.
https://github.com/python/cpython/pull/31717/files

Comment (vstinner) [reply Tools/gdb/libpython.py]: Oh I didn't know. Well, feel free to remove that ``u`` prefix in a separated PR :-) My PR doesn't add it at least :-)

**Pull request merged**